File: ios/MoltMarketplace/Core/Navigation/Route.swift
import Foundation

// MARK: - Route

/// Type-safe route definitions for all screens across milestones M0 through M4.
/// Routes that are not yet implemented render a placeholder view.
enum Route: Hashable {
    // MARK: - Home Tab

    case home
    case productDetail(productId: String)
    case productSearch
    case vendorStore(vendorId: String)
    case productReviews(productId: String)
    case writeReview(productId: String)

    // MARK: - Categories Tab

    case categories
    case categoryProducts(categoryId: String, categoryName: String)
    case productList(categoryId: String?, query: String?)

    // MARK: - Cart Tab

    case cart
    case checkout
    case checkoutAddress
    case checkoutShipping
    case checkoutPayment
    case orderConfirmation(orderId: String)

    // MARK: - Profile Tab

    case profile
    case orderList
    case orderDetail(orderId: String)
    case settings
    case addressManagement
    case wishlist
    case paymentMethods
    case notifications
    case recentlyViewed
    case priceAlerts

    // MARK: - Auth (Modal / Fullscreen)

    case login(returnTo: String?)
    case register
    case forgotPassword

    // MARK: - Onboarding (Fullscreen)

    case onboarding

    // MARK: - Auth Requirement

    /// Whether this route requires the user to be authenticated.
    var requiresAuth: Bool {
        switch self {
        case .checkout,
             .checkoutAddress,
             .checkoutShipping,
             .checkoutPayment,
             .orderConfirmation,
             .orderList,
             .orderDetail,
             .settings,
             .addressManagement,
             .wishlist,
             .paymentMethods,
             .notifications,
             .writeReview,
             .priceAlerts:
            return true
        case .home,
             .productDetail,
             .productSearch,
             .vendorStore,
             .productReviews,
             .categories,
             .categoryProducts,
             .productList,
             .cart,
             .profile,
             .login,
             .register,
             .forgotPassword,
             .onboarding,
             .recentlyViewed:
            return false
        }
    }

    // MARK: - Display Title

    /// Human-readable title for placeholder screens.
    var title: String {
        switch self {
        case .home: return String(localized: "nav_tab_home")
        case .categories: return String(localized: "nav_tab_categories")
        case .cart: return String(localized: "nav_tab_cart")
        case .profile: return String(localized: "nav_tab_profile")
        case .productDetail: return "Product Detail"
        case .productSearch: return "Search"
        case .vendorStore: return "Vendor Store"
        case .productReviews: return "Reviews"
        case .writeReview: return "Write Review"
        case .categoryProducts(_, let name): return name.isEmpty ? "Category" : name
        case .productList: return "Products"
        case .checkout: return "Checkout"
        case .checkoutAddress: return "Address"
        case .checkoutShipping: return "Shipping"
        case .checkoutPayment: return "Payment"
        case .orderConfirmation: return "Order Confirmed"
        case .orderList: return "Orders"
        case .orderDetail: return "Order Detail"
        case .settings: return "Settings"
        case .addressManagement: return "Addresses"
        case .wishlist: return "Wishlist"
        case .paymentMethods: return "Payment Methods"
        case .notifications: return "Notifications"
        case .recentlyViewed: return "Recently Viewed"
        case .priceAlerts: return "Price Alerts"
        case .login: return "Log In"
        case .register: return "Register"
        case .forgotPassword: return "Forgot Password"
        case .onboarding: return "Welcome"
        }
    }
}

File: android/app/src/main/java/com/molt/marketplace/core/navigation/Route.kt
package com.molt.marketplace.core.navigation

import kotlinx.serialization.Serializable

/**
 * Sealed interface defining all type-safe routes in the Molt Marketplace app.
 * Routes span milestones M0 through M4. Unimplemented routes render placeholder screens.
 */
sealed interface Route {

    // Home tab (M1-04)
    @Serializable
    data object Home : Route

    // Categories tab (M1-05)
    @Serializable
    data object Categories : Route

    @Serializable
    data class CategoryProducts(
        val categoryId: String,
        val categoryName: String,
    ) : Route

    @Serializable
    data class ProductList(
        val categoryId: String? = null,
        val query: String? = null,
    ) : Route

    // Shared (any tab)
    @Serializable
    data class ProductDetail(val productId: String) : Route

    @Serializable
    data object ProductSearch : Route

    @Serializable
    data class VendorStore(val vendorId: String) : Route

    @Serializable
    data class ProductReviews(val productId: String) : Route

    @Serializable
    data class WriteReview(val productId: String) : Route

    // Cart tab (M2)
    @Serializable
    data object Cart : Route

    @Serializable
    data object Checkout : Route

    @Serializable
    data object CheckoutAddress : Route

    @Serializable
    data object CheckoutShipping : Route

    @Serializable
    data object CheckoutPayment : Route

    @Serializable
    data class OrderConfirmation(val orderId: String) : Route

    // Profile tab (M3)
    @Serializable
    data object Profile : Route

    @Serializable
    data object OrderList : Route

    @Serializable
    data class OrderDetail(val orderId: String) : Route

    @Serializable
    data object Settings : Route

    @Serializable
    data object AddressManagement : Route

    @Serializable
    data object Wishlist : Route

    @Serializable
    data object PaymentMethods : Route

    @Serializable
    data object Notifications : Route

    @Serializable
    data object RecentlyViewed : Route

    @Serializable
    data object PriceAlerts : Route

    // Auth (modal / fullscreen)
    @Serializable
    data class Login(val returnTo: String? = null) : Route

    @Serializable
    data object Register : Route

    @Serializable
    data object ForgotPassword : Route

    // Onboarding (fullscreen)
    @Serializable
    data object Onboarding : Route
}

val Route.isAuthRequired: Boolean
    get() = when (this) {
        is Route.Checkout,
        is Route.CheckoutAddress,
        is Route.CheckoutShipping,
        is Route.CheckoutPayment,
        is Route.OrderConfirmation,
        is Route.OrderList,
        is Route.OrderDetail,
        is Route.Settings,
        is Route.AddressManagement,
        is Route.Wishlist,
        is Route.PaymentMethods,
        is Route.Notifications,
        is Route.WriteReview,
        is Route.PriceAlerts,
        -> true

        is Route.Home,
        is Route.Categories,
        is Route.CategoryProducts,
        is Route.ProductList,
        is Route.ProductDetail,
        is Route.ProductSearch,
        is Route.VendorStore,
        is Route.ProductReviews,
        is Route.Cart,
        is Route.Profile,
        is Route.Login,
        is Route.Register,
        is Route.ForgotPassword,
        is Route.Onboarding,
        is Route.RecentlyViewed,
        -> false
    }
